pow/src/main/java/tech/pegasys/artemis/pow/Eth1Manager.java

```diff
@@ -13,113 +13,121 @@
 import tech.pegasys.artemis.util.async.SafeFuture;
 import tech.pegasys.artemis.util.config.Constants;
 
-public class Eth1MinGenesisTimeBlockFinder {
+public class Eth1Manager {
 
   private static final Logger LOG = LogManager.getLogger();
 
   private final Eth1Provider eth1Provider;
-  private final MinGenesisTimeBlockEventChannel minGenesisTimeBlockEventChannel;
   private final AsyncRunner asyncRunner;
+  private final Eth1EventsChannel eth1EventsChannel;
+  private final DepositProcessingController depositProcessingController;
 
-  private volatile Disposable latestBlockDisposable;
-
-  public Eth1MinGenesisTimeBlockFinder(
+  public Eth1Manager(
       Eth1Provider eth1Provider,
-      MinGenesisTimeBlockEventChannel minGenesisTimeBlockEventChannel,
-      AsyncRunner asyncRunner) {
+      AsyncRunner asyncRunner,
+      Eth1EventsChannel eth1EventsChannel,
+      DepositProcessingController depositProcessingController) {
     this.eth1Provider = eth1Provider;
-    this.minGenesisTimeBlockEventChannel = minGenesisTimeBlockEventChannel;
     this.asyncRunner = asyncRunner;
+    this.eth1EventsChannel = eth1EventsChannel;
+    this.depositProcessingController = depositProcessingController;
   }
 
   public void start() {
-    findAndPublishFirstValidBlock()
-        .finish(
-            () ->
-                LOG.info(
-                    "Eth1MinGenesisBlockFinder successfully found first "
-                        + "(time) valid genesis block"));
+
+    SafeFuture<EthBlock.Block> headBlock = getHead();
+
+    SafeFuture<Boolean> isHeadBeforeMinGenesis = isHeadBeforeMinGenesis(headBlock);
+
+    isHeadBeforeMinGenesis.thenAccept(
+        bool -> {
+          if (bool) headBeforeMinGenesisMode(headBlock);
+          else headAfterMinGenesisMode(headBlock);
+        });
+  }
+
+  public SafeFuture<Void> headBeforeMinGenesisMode(SafeFuture<EthBlock.Block> head) {
+    return head.thenCompose(
+            headBlock ->
+                depositProcessingController.fetchDepositsFromGenesisTo(headBlock.getNumber()))
+        .thenRun(
+            () -> {
+              EthBlock.Block block = checkNotNull(head.getNow(null));
+              depositProcessingController.switchToBlockByBlockMode();
+              depositProcessingController.startSubscription(block.getNumber());
+            });
+  }
+
+  public SafeFuture<Void> headAfterMinGenesisMode(SafeFuture<EthBlock.Block> head) {
+    SafeFuture<EthBlock.Block> minGenesisBlock =
+        head.thenCompose(this::findMinGenesisTimeBlockInHistory);
+
+    return minGenesisBlock
+        .thenCompose(
+            block -> depositProcessingController.fetchDepositsFromGenesisTo(block.getNumber()))
+        .thenRun(
+            () -> {
+              EthBlock.Block block = checkNotNull(minGenesisBlock.getNow(null));
+              postMinGenesisTimeBlock(eth1EventsChannel, block);
+              depositProcessingController.startSubscription(block.getNumber());
+            });
```

Comment:
Using chaining is probably better here than `getNow`. 

```
public SafeFuture<Void> headAfterMinGenesisMode(EthBlock.Block head) {
    return findMinGenesisTimeBlockInHistory(head)
        .thenCompose(this::sendDepositsUpToMinGenesis)
        .thenAccept(
            minGenesisTimeBlock -> {
              postMinGenesisTimeBlock(eth1EventsChannel, minGenesisTimeBlock);
              depositProcessingController.startSubscription(minGenesisTimeBlock.getNumber());
            });
  }

  private SafeFuture<Block> sendDepositsUpToMinGenesis(final Block minGenesisTimeBlock) {
    return depositProcessingController
        .fetchDepositsFromGenesisTo(minGenesisTimeBlock.getNumber())
        .thenApply(__ -> minGenesisTimeBlock);
  }
```

Reply:
yeah, this is cleaner, although its a bit vague (from sendDepositsUpToMinGenesis's own pov) why it returns the minGenesisBlock. done tho.